Services/DeviceModels.cs

```diff
@@ -29,49 +51,61 @@ public class DeviceModels : IDeviceModels
 
         private readonly IServicesConfig config;
         private readonly ILogger log;
-
-        private List<string> deviceModelFiles;
-        private List<DeviceModel> deviceModels;
+        private readonly IStorageAdapterClient storage;
+        
+        private CustomDeviceModels customDeviceModels;
+        private StockDeviceModels stockDeviceModels;
 
         public DeviceModels(
+            IStorageAdapterClient storage,
+            ICustomDeviceModels CustomDeviceModels,
+            IStockDeviceModels StockDeviceModels,
             IServicesConfig config,
             ILogger logger)
         {
+            this.storage = storage;
             this.config = config;
             this.log = logger;
-            this.deviceModelFiles = null;
-            this.deviceModels = null;
+            this.stockDeviceModels = new StockDeviceModels(
+                this.config,
+                this.log);
+            this.customDeviceModels = new CustomDeviceModels(
+                this.storage,
+                this.log);
         }
 
-        public IEnumerable<DeviceModel> GetList()
+        /// <summary>
+        /// Get list of device models.
+        /// </summary>
+        public async Task<IEnumerable<DeviceModel>> GetListAsync()
         {
-            if (this.deviceModels != null) return this.deviceModels;
-
-            this.deviceModels = new List<DeviceModel>();
+            var deviceModels = new List<DeviceModel>();
 
             try
             {
-                var files = this.GetDeviceModelFiles();
-                foreach (var f in files)
-                {
-                    var c = JsonConvert.DeserializeObject<DeviceModel>(File.ReadAllText(f));
-                    this.deviceModels.Add(c);
-                }
+                var stockDeviceModels = this.stockDeviceModels.GetList();
+                var customDeviceModels = await this.customDeviceModels.GetListAsync();
+                deviceModels = stockDeviceModels
+                    .Concat(customDeviceModels)
+                    .ToList();
             }
             catch (Exception e)
             {
-                this.log.Error("Unable to load Device Model configuration",
+                this.log.Error("Unable to load Device Model ",
                     () => new { e.Message, Exception = e });
 
-                throw new InvalidConfigurationException("Unable to load Device Model configuration: " + e.Message, e);
+                throw new Exception("Unable to load Device Model : " + e.Message, e);
```

Comment:
typo

Reply:
you don't need to concatenate the error message. Also better to use ExternalDependencyException:
```
throw new ExternalDependencyException("Unable to load device models", e)
```

Reply:
Done.